Frameworks/Foundation/NSOperation.mm

```diff
@@ -111,32 +115,36 @@ - (id)init {
  @Status Interoperable
 */
 - (BOOL)isReady {
-    std::lock_guard<std::recursive_mutex> lock(_readyLock);
+    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
     return _ready;
 }
 
 /**
  @Status Interoperable
 */
 - (void)addDependency:(id)operation {
-    std::lock_guard<std::recursive_mutex> lock(_dependenciesLock);
+    std::lock_guard<std::recursive_mutex> dependenciesLock(_dependenciesLock);
     [self willChangeValueForKey:@"dependencies"];
     [_dependencies addObject:operation];
-    [operation addObserver:self forKeyPath:@"isFinished" options:0 context:(void*)NSOperationContext];
+    [operation addObserver:self forKeyPath:@"isFinished" options:NSKeyValueObservingOptionInitial context:(void*)NSOperationContext];
     [self didChangeValueForKey:@"dependencies"];
+    _outstandingDependenciesCount++; // will be decremented if necessary in observeValueForKey, due to NSKeyValueObservingOptionInitial
     [self _checkReady];
```

Comment:
Possibly reorder these, so that the increment always stages before the decrement.
Otherwise, you might run into an underflow situation when addObserver drops it below 0.